network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -78,65 +79,37 @@ func AuthorizedSenderValidator(log zerolog.Logger, channel network.Channel, getI
 	}
 }
 
-// isAuthorizedSender checks if node is an authorized role and is not ejected.
-func isAuthorizedSender(identity *flow.Identity, channel network.Channel, code uint8) error {
-	// get authorized roles list
-	roles, err := getRoles(channel, code)
-	if err != nil {
-		return err
-	}
-
-	if !roles.Contains(identity.Role) {
-		return fmt.Errorf("sender is not authorized to send this message type")
-	}
-
-	return nil
-}
-
-// getRoles returns list of authorized roles for the channel associated with the message code provided
-func getRoles(channel network.Channel, msgTypeCode uint8) (flow.RoleList, error) {
+// isAuthorizedSender checks if node is an authorized role.
+func isAuthorizedSender(identity *flow.Identity, channel network.Channel, code network.MessageCode) error {
 	// echo messages can be sent by anyone
-	if msgTypeCode == cborcodec.CodeEcho {
-		return flow.Roles(), nil
+	if code.Code == codec.CodeEcho {
+		return nil
 	}
 
-	// get message type codes for all messages communicated on the channel
-	codes, ok := getCodes(channel)
-	if !ok {
-		return nil, fmt.Errorf("could not get message codes for unknown channel: %s", channel)
+	var authorizedRolesByChannel flow.RoleList
+
+	// handle cluster prefixed channels and check and get authorized roles list
+	if prefix, ok := network.ClusterChannelPrefix(channel); ok {
+		authorizedRolesByChannel = code.AuthorizedRolesByChannel(network.Channel(prefix))
+	} else {
+		authorizedRolesByChannel = code.AuthorizedRolesByChannel(channel)
 	}
 
-	// check if message type code is in list of codes corresponding to channel
-	if !containsCode(codes, msgTypeCode) {
-		return nil, fmt.Errorf("invalid message type being sent on channel")
+	// check if role is authorized to send message on channel
```

Comment:
Here we check the message code authorization. Check if the current channel is in the message code's authorizedChannelsMap (asserts that this messsage is allowed to be sent on the channel) , then check if the senders role is in the channels authorized role list (asserts the sender is allowed to send this message on the current channel)